libtsuba/src/RDGTopology.cpp

```diff
@@ -0,0 +1,579 @@
+#include "katana/URI.h"
+#include "katana/config.h"
+#include "tsuba/Errors.h"
+#include "tsuba/FaultTest.h"
+#include "tsuba/FileFrame.h"
+#include "tsuba/FileView.h"
+#include "tsuba/RDG.h"
+#include "tsuba/tsuba.h"
+
+namespace tsuba {
+
+std::string
+RDGTopology::path() const {
+  if (metadata_entry_valid()) {
+    return metadata_entry_->path_;
+  }
+
+  KATANA_LOG_WARN(
+      "trying to get topology path, but not linked to a metadata entry. "
+      "Returning empty string");
+  return std::string();
+}
+
+void
+RDGTopology::set_path(const std::string& path) {
+  KATANA_LOG_VASSERT(
+      metadata_entry_valid(),
+      "metadata_entry_->must be set before we can set the topology path");
+  metadata_entry_->path_ = path;
+}
+
+void
+RDGTopology::set_metadata_entry(PartitionTopologyMetadataEntry* entry) {
+  KATANA_LOG_ASSERT(entry != nullptr);
+  metadata_entry_ = entry;
+  KATANA_LOG_ASSERT(metadata_entry_valid());
+}
+
+bool
+RDGTopology::metadata_entry_valid() const {
+  return (metadata_entry_ != nullptr);
+}
+
+katana::Result<void>
+RDGTopology::Bind(const katana::Uri& metadata_dir) {
+  if (file_store_bound_) {
+    KATANA_LOG_WARN("topology already bound, nothing to do");
+    return katana::ResultSuccess();
+  }
+
+  katana::Uri t_path = metadata_dir.Join(path());
+  KATANA_LOG_DEBUG(
+      "binding to entire topology file at path {}", t_path.string());
+  if (auto res = file_storage_.Bind(t_path.string(), true); !res) {
+    return res.error();
+  }
+
+  file_store_bound_ = true;
+  storage_valid_ = true;
+
+  return katana::ResultSuccess();
+}
+
+katana::Result<void>
+RDGTopology::Bind(
+    const katana::Uri& metadata_dir, uint64_t begin, uint64_t end,
+    bool resolve) {
+  katana::Uri t_path = metadata_dir.Join(path());
+  KATANA_LOG_DEBUG(
+      "binding from {} to {} with topology file at path {}", begin, end,
+      t_path.string());
+  if (auto res = file_storage_.Bind(t_path.string(), begin, end, resolve);
+      !res) {
+    return res.error();
+  }
+
+  file_store_bound_ = true;
+  storage_valid_ = true;
+
+  return katana::ResultSuccess();
+}
+
+katana::Result<void>
+RDGTopology::Map() {
+  if (file_store_mapped_) {
+    return katana::ResultSuccess();
+  }
+
+  if (!file_store_bound_) {
+    return KATANA_ERROR(
+        ErrorCode::InvalidArgument,
+        "topology must be bound before it is mapped");
+  }
+
+  const auto* data = file_storage_.ptr<uint64_t>();
+
+  size_t min_size = 4;
+  if (file_storage_.size() < min_size) {
+    return katana::ErrorCode::InvalidArgument;
+  }
+
+  if (data[0] != 1) {
+    return katana::ErrorCode::InvalidArgument;
+  }
+
+  // ensure the data file matches the metadata
+  KATANA_LOG_VASSERT(
+      num_nodes_ == data[2], "expected {} nodes, found {} nodes", num_nodes_,
+      data[2]);
+  KATANA_LOG_VASSERT(
+      num_edges_ == data[3], "expected {} edges, found {} edges", num_edges_,
+      data[3]);
+
+  adj_indices_ = &data[4];
+  const uint64_t* cursor = adj_indices_;
```

Comment:
This code with `cursor` and its math is a bit low-level and error prone. My recommendation is that we introduce a "byte data iterator". This iterator supports usual input iterator stuff, as well as things like `AdvanceBy<T>()`, which advances the cursor inside the iterator by `sizeof(T)`. All this math of `/2 + %2` will be hidden in the iterator and life will be better :).

Reply:
Fully in support of this suggestion, but it might have to wait for a future PR.

Reply:
Add a TODO please (assuming you've decided to postpone this).